src/alloc/allocext/linux.rs

```diff
@@ -0,0 +1,131 @@
+extern crate std;
+use self::std::process::abort;
+use crate::{alloc::*, Prot};
+use core::mem::{self, size_of};
+use core::ptr::{self, NonNull};
+use core::slice;
+
+use self::memfd_secret_alloc::*;
+
+mod memfd_secret_alloc {
+    use core::convert::TryInto;
+    use std::{io, println};
+
+    use super::*;
+
+    #[inline]
+    pub unsafe fn alloc_memfd_secret(size: usize) -> Option<(NonNull<u8>, libc::c_int)> {
+        let fd: Result<libc::c_int, _> = libc::syscall(libc::SYS_memfd_secret, 0).try_into();
+
+        let fd = fd.ok().filter(|&fd| fd >= 0)?;
+
+        // File size is set using ftruncate
+        let _ = libc::ftruncate(fd, size as libc::off_t);
+
+        let ptr = libc::mmap(
+            ptr::null_mut(),
+            size,
+            Prot::ReadWrite,
+            libc::MAP_SHARED,
+            fd,
+            0,
+        );
+
+        if ptr == libc::MAP_FAILED {
+            return None;
+        }
+
+        NonNull::new(ptr as *mut u8).map(|ptr| (ptr, fd))
+    }
+}
+
+unsafe fn _memfd_secret(size: usize) -> Option<*mut u8> {
+    ALLOC_INIT.call_once(|| alloc_init());
+
+    if size >= ::core::usize::MAX - PAGE_SIZE * 4 {
+        return None;
+    }
+
+    // aligned alloc ptr
+    let size_with_canary = CANARY_SIZE + size;
+    let unprotected_size = page_round(size_with_canary);
+    let total_size = PAGE_SIZE + PAGE_SIZE + unprotected_size + PAGE_SIZE;
+    let (base_ptr, fd) = alloc_memfd_secret(total_size)?;
+    let base_ptr = base_ptr.as_ptr();
+    let fd_ptr = base_ptr.add(size_of::<usize>());
```

Comment:
maybe we should add an assert size + fd <= page

Reply:
Thank you, made the change on line 43. Please let me know if that wasn't what you had in mind.